redis/src/main/scala/zio/redis/TestExecutor.scala

```diff
@@ -1,33 +1,65 @@
 package zio.redis
 
+import scala.annotation.tailrec
 import scala.collection.compat.immutable.LazyList
+import scala.util.Try
 
 import zio._
+import zio.duration._
 import zio.redis.RedisError.ProtocolError
 import zio.redis.RespValue.BulkString
-import zio.stm.{ STM, TMap, TRef, USTM }
+import zio.stm.{ random => _, _ }
 
 private[redis] final class TestExecutor private (
+  lists: TMap[String, Vector[String]],
   sets: TMap[String, Set[String]],
   strings: TMap[String, String],
   randomPick: Int => USTM[Int],
-  hyperLogLogs: TMap[String, Seq[String]]
+  hyperLogLogs: TMap[String, Set[String]]
 ) extends RedisExecutor.Service {
 
   def execute(command: Chunk[RespValue.BulkString]): IO[RedisError, RespValue] =
     for {
-      name   <- ZIO.fromOption(command.headOption).orElseFail(ProtocolError("Malformed command."))
-      result <- runCommand(name.asString, command.tail).commit
+      name <- ZIO.fromOption(command.headOption).orElseFail(ProtocolError("Malformed command."))
+      result <- name.asString match {
+                  case api.Lists.BlPop.name =>
+                    val timeout = command.tail.last.asString.toInt
+                    runBlockingCommand(name.asString, command.tail, timeout, RespValue.NullArray)
+
+                  case api.Lists.BrPop.name =>
+                    val timeout = command.tail.last.asString.toInt
+                    runBlockingCommand(name.asString, command.tail, timeout, RespValue.NullArray)
+
+                  case api.Lists.BrPopLPush.name =>
+                    val timeout = command.tail.last.asString.toInt
+                    runBlockingCommand(name.asString, command.tail, timeout, RespValue.NullBulkString)
+
+                  case _ => runCommand(name.asString, command.tail).commit
+                }
     } yield result
 
+  private def runBlockingCommand(
+    name: String,
+    input: Chunk[RespValue.BulkString],
+    timeout: Int,
+    respValue: RespValue
+  ): UIO[RespValue] =
+    if (timeout > 0)
+      runCommand(name, input).commit
+        .timeout(timeout.seconds)
+        .map(_.getOrElse(respValue))
+        .provideLayer(zio.clock.Clock.live)
+    else
+      runCommand(name, input).commit
+
   private def errResponse(cmd: String): RespValue.BulkString =
     RespValue.bulkString(s"(error) ERR wrong number of arguments for '$cmd' command")
 
   private def onConnection(command: String, input: Chunk[RespValue.BulkString])(
     res: => RespValue.BulkString
   ): USTM[BulkString] = STM.succeedNow(if (input.isEmpty) errResponse(command) else res)
 
-  private[this] def runCommand(name: String, input: Chunk[RespValue.BulkString]): STM[RedisError, RespValue] = {
+  private[this] def runCommand(name: String, input: Chunk[RespValue.BulkString]): USTM[RespValue] = {
```

Comment:
We should make sure no unsafe calls are made, e.g. head on potentially empty chunk. It can (and should) be done in a separate pull request.

Reply:
I am not sure if I understand what you are trying to say here. From my point of view, we statically know which inputs we feed into each command and therefore we are able to decode the input even with the use of unsafe operations like `head` etc. How do you imagine that we ensure that `no unsafe calls are made`?

Reply:
What I meant is double-checking if we know statically (last time I took a look there were quite a few head calls), and if now, changing the type to NonEmptyChunk :). If there's nothing to be done here, it's a no-op (even better :) ).

Reply:
Alright got it. But as you said this can be done in a follow-up PR.